src/sparse_blas/backends/cusparse/cusparse_task.hpp

```diff
@@ -0,0 +1,382 @@
+    cgh.require(workspace_placeholder_acc);
+    cgh.host_task([functor, queue, workspace_placeholder_acc,
+                   capture_only_accessors...](sycl::interop_handle ih) {
+        auto unused = std::make_tuple(capture_only_accessors...);
+        (void)unused;
+        auto sc = CusparseScopedContextHandler(queue, ih);
+        functor(sc, workspace_placeholder_acc);
+    });
+}
+
+template <typename Functor, typename... CaptureOnlyAcc>
+void submit_native_command_ext(sycl::handler &cgh, sycl::queue &queue, Functor functor,
+                               const std::vector<sycl::event> &dependencies,
+                               CaptureOnlyAcc... capture_only_accessors) {
+    // Only capture the accessors to ensure the dependencies are properly handled
+    // The accessors's pointer have already been set to the native container types in previous functions
+#ifdef SYCL_EXT_ONEAPI_ENQUEUE_NATIVE_COMMAND
+    cgh.ext_codeplay_enqueue_native_command(
+        [functor, queue, dependencies, capture_only_accessors...](sycl::interop_handle ih) {
+            auto unused = std::make_tuple(capture_only_accessors...);
+            (void)unused;
+            auto sc = CusparseScopedContextHandler(queue, ih);
+            // The functor using ext_codeplay_enqueue_native_command need to
+            // explicitly wait on the events for the SPARSE domain. The
+            // extension ext_codeplay_enqueue_native_command is used to launch
+            // the compute operation which depends on the previous optimize
+            // step. In cuSPARSE the optimize step is synchronous but it is
+            // asynchronous in oneMKL Interface. The optimize step may not use
+            // the CUDA stream which would make it impossible for
+            // ext_codeplay_enqueue_native_command to automatically ensure it
+            // has completed before the compute function starts. These waits are
+            // used to ensure the optimize step has completed before starting
+            // the computation.
+            for (auto event : dependencies) {
+                event.wait();
+            }
+            functor(sc);
+        });
+#else
+    (void)dependencies;
+    submit_host_task(cgh, queue, functor, capture_only_accessors...);
+#endif
+}
+
+template <typename Functor, typename... CaptureOnlyAcc>
+void submit_native_command_ext_with_acc(sycl::handler &cgh, sycl::queue &queue, Functor functor,
+                                        const std::vector<sycl::event> &dependencies,
+                                        sycl::accessor<std::uint8_t> workspace_placeholder_acc,
+                                        CaptureOnlyAcc... capture_only_accessors) {
+    // Only capture the accessors to ensure the dependencies are properly handled
+    // The accessors's pointer have already been set to the native container types in previous functions
```

Comment:
Do you know if there is a SYCL Specification guarantee that the pointers underlying `sycl::buffer`s never change for the life of the `sycl::buffer`? Particularly, if data is moved from the device to host and then back, for instance, is the pointer that existed before the data movement guaranteed to be the same as that after the movement? Only then would it be fine to only capture accessors without resetting the pointers, isn't it?

Reply:
This is a good question I discussed this issue before with Gordon Brown. There are no guarantee in the SYCL specification unfortunately but we agreed that in practice the implementations have no reason to change the underlying pointer. Even if the data is copied to the host, the data must be valid on the device as long as the buffer is alive (which we ensure it is).
I'm not sure if there is any plan to clarify this guarantee in the spec given that there were discussions to deprecate the SYCL buffers at the time. If this is a concern to you I can raise this issue again.

Reply:
No need to raise this issue again, but maybe just expand the comment with an additional sentence stating the assumption about underlying pointers remaining unchanged in SYCL implementations.

Reply:
This is clarified in https://github.com/oneapi-src/oneMKL/pull/527/commits/050f22af0410d53bcc95eb97080c8414dc10f997

Reply:
Great, thanks!